src/main/java/pro/sky/Examiner/service/JavaQuestionService.java

```diff
@@ -30,21 +31,37 @@ public Question add(Question question) {
 
     @Override
     public Question remove(Question question) {
-        questions.remove(question);
-        return question;
+        if (questions.remove(question)) {
+            return question;
+        }
+        return null;
     }
 
     @Override
     public Set<Question> getAll() {
-        return questions;
+//        return questions;//forTESTS___  getAllCopy()
+        return Collections.unmodifiableSet(questions);
     }
 
     @Override
     public Question getRandomQuestion() {
-        List<Question> listQuestion = new ArrayList<>(questions);
-        Random random = new Random();
-        int r = random.nextInt(questions.size());
-        return listQuestion.get(r);
+
+        if (questions.isEmpty()) {
+        throw new NoQuestionsException();
+        }
+
+        int index = random.nextInt(questions.size());
+        var it = questions.iterator();
+        //если вопросов не будет, то nextInt кинет исключение - хорошо бы сделать проверку
+        int i = 0;
+        while (it.hasNext()) {
+            Question q = it.next();
+            if (i == index) {
+                return q;
+            }
+         i++;
+        }
+        return null;// или вернуть null
```

Comment:
```suggestion
       return new ArrayList<>(questions.size()).get(random.nextInt(questions.size()));
```
и всё, 1 строка всего лишь)